### PR TITLE
Add Support for MySQL

### PR DESCRIPTION
**Pull request by TasdidurRahman**

### CODE REVIEW

**Comment by TasdidurRahman** [comment]
Add Support for MySQL

File: apis/schema/v1alpha1/mysqldatabase_types.go [hunk 1]
```diff
@@ -28,18 +32,28 @@ type MySQLDatabaseSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of MySQLDatabase. Edit mysqldatabase_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	MySQLRef          kmodulesv1alpha1.AppReference `json:"my_sql_ref"`
```
Comment (heheh13) [inline apis/schema/v1alpha1/mysqldatabase_types.go:35]: Add comment before every field describing what's its doing.

Comment (heheh13) [reply apis/schema/v1alpha1/mysqldatabase_types.go:35]: like 
// MySQLRef is the reference to the origin database instance.

File: controllers/schema/mysqldatabase_controller.go [hunk 2]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		if curSecacreq.Status.Secret == nil {
+			//secret is not patched yet
+			klog.Errorf("\nWaiting for the secret to be patched with SAR %s\n", curSecacreq.Name)
+			//return with nil error : don't panic , we're watching the request
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// check if we have run the init script
+	if schemaobj.Spec.Init.Script != (core.VolumeSource{}) && schemaobj.Spec.Init.Initialized != true {
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller.go:318]: https://github.com/kubedb/mysql/blob/master/pkg/controller/mysql.go#L112

File: apis/schema/v1alpha1/mysqldatabase_types.go [hunk 3]
```diff
@@ -28,18 +32,28 @@ type MySQLDatabaseSpec struct {
...
+	VaultRef          kmodulesv1.ObjectReference    `json:"vault_ref"`
+	DatabaseConfig    Database                      `json:"database_config"`
+	Subjects          []rbac.Subject                `json:"subjects"`
+	Init              Initer                        `json:"init,omitempty"`
```
Comment (heheh13) [inline apis/schema/v1alpha1/mysqldatabase_types.go:39]: I think `InitSpec` looks fine than Initer
seems like init spec should be optional. If it then add 
+optional
Init *InitSpec

File: apis/schema/v1alpha1/mysqldatabase_types.go [hunk 4]
```diff
@@ -28,18 +32,28 @@ type MySQLDatabaseSpec struct {
...
+	RepoRef           RestoreRef                    `json:"repo_ref,omitempty"`
```
Comment (heheh13) [inline apis/schema/v1alpha1/mysqldatabase_types.go:40]: seems another optional field.
```
+optional
Restore           RestoreConf                   `json:"restore,omitempty"`
```

File: apis/schema/v1alpha1/mysqldatabase_types.go [hunk 5]
```diff
@@ -62,3 +76,52 @@ type MySQLDatabaseList struct {
...
+const (
+	TerminationPolicyDelete         TerminationPolicy = "Delete"
+	TerminationPolicyDoNotTerminate TerminationPolicy = "DoNotTerminate"
+
+	//readOnly
+	enable  readOnly = 1
+	disable readOnly = 0
+)
+
+type MySQLDatabasePhase string
```
Comment (heheh13) [inline apis/schema/v1alpha1/mysqldatabase_types.go:112]: Remove if not used.

File: controllers/schema/mysqldatabase_controller.go [hunk 6]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		&kvenginev1a1.SecretAccessRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      schemaobj.GetSecretAccReqName(),
+				Namespace: schemaobj.GetSecretAccReqNamespace(),
+			},
+		},
+		func(obj client.Object, createOp bool) client.Object {
+			req := obj.(*kvenginev1a1.SecretAccessRequest)
+			req.Spec.RoleRef = core.TypedLocalObjectReference{
+				Kind: "MySQLRole",
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller.go:277]: use constant

File: controllers/schema/mysqldatabase_controller.go [hunk 7]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+			return req
+		})
+	if err != nil {
+		klog.Errorf("\nError createOrPatch secretAccessRequest %s\n", schemaobj.GetSecretAccReqName())
+		return ctrl.Result{}, err
+	}
+	fmt.Printf("%s %s\n", secretAccessRequest.GetName(), vt)
+
+	//get the current secret access request
+	curSecacreq, err := r.GetSecAccReqforSchema(ctx, schemaobj)
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller.go:296]: `curSecretAccessRequest`

File: controllers/schema/mysqldatabase_controller.go [hunk 8]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		// init script is yet to run
+		// get the vault secret : script will be executed from vault user account
+		vaultsecret, err := r.GetSecretforSAR(ctx, schemaobj)
+		if err != nil {
+			klog.Errorf("\nError fetching secret for secret access request %s\n", curSecacreq.Name)
+			return ctrl.Result{}, err
+		}
+		if vaultsecret == nil {
+			klog.Errorf("\nError while fetching the vault secret for request %s\n", curSecacreq.Name)
+			// todo handle the case more precisely
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller.go:328]: .

File: controllers/schema/mysqldatabase_controller.go [hunk 9]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+					Namespace: schemaobj.GetJobNamespace(),
+				},
+			},
+			func(obj client.Object, createOp bool) client.Object {
+				jb := obj.(*batch.Job)
+
+				jb.Spec.Template.Spec.Containers = kcore.UpsertContainers(jb.Spec.Template.Spec.Containers, []core.Container{
+					{
+						Name:         schemaobj.Name + "-job-container",
+						Image:        "mysql:" + mysqlobj.Spec.Version, //"mysql:8.0.27",//getMysqlVersion(c, dbobj.Spec.MySQLRef),
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller.go:345]: get from mysqlverisons crd

File: controllers/schema/mysqldatabase_controller.go [hunk 10]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		For(&schemav1a1.MySQLDatabase{}).
+		Owns(&kvenginev1a1.MySQLRole{}).
+		Owns(&kvenginev1a1.SecretEngine{}).
+		Owns(&kvenginev1a1.SecretAccessRequest{}).
+		Watches(&source.Kind{Type: &kvaultv1a1.VaultServer{}}, vaultHandler).
+		Watches(&source.Kind{Type: &api.MySQL{}}, mysqlHandler).
 		Complete(r)
 }
+
+//trash :
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller.go:519]: Remove Trash :smile:

File: controllers/schema/mysqldatabase_controller_helper.go [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+	if err != nil {
+		klog.Errorf("\nError creating restoresession\n")
+		return kutil.VerbUnchanged, err
+	}
+	fmt.Printf("%s %s\n", restoreses.GetName(), vt)
+
+	return vt, nil
+}
+
+func CreateLocalRepo(r *MySQLDatabaseReconciler, schemaobj *schemav1a1.MySQLDatabase) (kutil.VerbType, error) {
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller_helper.go:188]: `CreateLocalRepository`

File: controllers/schema/mysqldatabase_controller_helper.go [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+		Name:      objref.Name,
+	}
+	err = r.Client.Get(ctx, key, vaultsecret)
+	if err != nil {
+		return nil, err
+	}
+	return vaultsecret, err
+}
+
+func (r *MySQLDatabaseReconciler) GetSecAccReqforSchema(ctx context.Context, schemaobj *schemav1a1.MySQLDatabase) (*kvenginev1a1.SecretAccessRequest, error) {
```
Comment (heheh13) [inline controllers/schema/mysqldatabase_controller_helper.go:302]: .

**Review by heheh13 — CHANGES_REQUESTED** [review]
- fix typos
- remove unnecessary Comments
- use log instead print statements
- use camelCase
- recheck the function names. 
- run linter.`make fmt lint`

File: apis/schema/v1alpha1/mysqldatabase_types.go (hunk 1 above)
Comment (TasdidurRahman) [reply apis/schema/v1alpha1/mysqldatabase_types.go:35]: add descriptive comments 
change snake_case to camelCase

File: apis/schema/v1alpha1/mysqldatabase_types.go (hunk 3 above)
Comment (TasdidurRahman) [reply apis/schema/v1alpha1/mysqldatabase_types.go:39]: change ```Init Initer``` to ```InitSpec *InitSpec```

File: apis/schema/v1alpha1/mysqldatabase_types.go (hunk 4 above)
Comment (TasdidurRahman) [reply apis/schema/v1alpha1/mysqldatabase_types.go:40]: change ```RepoRef  RestoreRef ``` to ```Restore  *RestoreConf```

Comment (TasdidurRahman) [reply apis/schema/v1alpha1/mysqldatabase_types.go:40]: add ```optional``` marker

File: apis/schema/v1alpha1/mysqldatabase_types.go (hunk 5 above)
Comment (TasdidurRahman) [reply apis/schema/v1alpha1/mysqldatabase_types.go:112]: remove unused constants

File: controllers/schema/mysqldatabase_controller.go [hunk 13]
```diff
@@ -18,19 +18,37 @@ package schema
...
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
 // MySQLDatabaseReconciler reconciles a MySQLDatabase object
 type MySQLDatabaseReconciler struct {
 	client.Client
-	Scheme *runtime.Scheme
+	Scheme           *runtime.Scheme
+	AppCatagolClient appcat_cs.Interface
```
[the remark this answers is not in the document]
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:50]: fix typo

File: controllers/schema/mysqldatabase_controller.go [hunk 14]
```diff
@@ -18,19 +18,37 @@ package schema
...
+	//Recorder         record.EventRecorder
```
[the remark this answers is not in the document]
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:51]: remove recorder

File: controllers/schema/mysqldatabase_controller.go [hunk 15]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	// your logic here
```
[the remark this answers is not in the document]
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:70]: remove comment

File: controllers/schema/mysqldatabase_controller.go [hunk 16]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+			klog.Errorf("\nMySQLDatabase %s does not exist anymore!\n", req.Name) //object deleted
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("\nError fetching MySQLDatabase object %s\n", req.Name) // object not deleted but error in fetching
+		return ctrl.Result{}, err
+	}
+
+	klog.Info("\nSTARTED RECONCILER FOR : ", schemaobj.Name, "\n")
+
+	//prepare finalizer to add later
```
[the remark this answers is not in the document]
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:86]: remove comment

File: controllers/schema/mysqldatabase_controller.go [hunk 17]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+	if schemaobj.Status.Phase != DatabaseCreated {
+		//target database is not created
+		//create the database
+		if err = DatabaseOperation(schemaobj, r, ctx, CreateDatabase); err != nil {
+			//error while creating database
+			klog.Errorf("\nError while creating database %s\n", schemaobj.Spec.DatabaseConfig.Name)
+			return ctrl.Result{}, err
+		}
+		schemaobj.Status.Phase = DatabaseCreated
+		_, _, errr := CreateOrPatchStatus(r.Client, &schemav1a1.MySQLDatabase{
```
[the remark this answers is not in the document]
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:194]: change err to err2

File: controllers/schema/mysqldatabase_controller.go (hunk 6 above)
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:277]: create constant 
replace with constant

File: controllers/schema/mysqldatabase_controller.go [hunk 18]
```diff
@@ -51,12 +69,579 @@ func (r *MySQLDatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+	fmt.Printf("%s %s\n", secretAccessRequest.GetName(), vt)
+
+	//get the current secret access request
+	curSecacreq, err := r.GetSecAccReqforSchema(ctx, schemaobj)
+	if err != nil {
+		//got error while fetching the current secret access request
+		klog.Errorf("\nError fetching SecretAccessRequest %s\n", schemaobj.GetSecretAccReqName())
+		return ctrl.Result{}, err
+	} else {
+		//chech if the request is being approved
```
[the remark this answers is not in the document]
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:302]: fix typo

File: controllers/schema/mysqldatabase_controller.go (hunk 2 above)
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:318]: change ```schemaobj.Spec.Init.Script != (core.VolumeSource{})``` to ```schemaobj.Spec.InitSpec != nil```

File: controllers/schema/mysqldatabase_controller.go (hunk 10 above)
Comment (TasdidurRahman) [reply controllers/schema/mysqldatabase_controller.go:519]: remove trash

**Comment by TasdidurRahman** [comment]
https://github.com/kubedb/schema-manager/pull/8

**Pull request closed without merging**